docs/src/index.md

```diff
@@ -4,6 +4,349 @@ DocTestSetup = :(using ChainRulesCore, ChainRules)
+```
+
+ -------------------------------
+
+## On writing good `rrule` / `frule` methods
+
+### Return Zero or One
+
+Rather than `0` or `1` or even rather than `zeros(n)`, `ones(n)`, or the identity matrix `I`.
+
+### Use `Thunk`s appropriately:
+
+If work is only required for one of the returned differentials, then it should be wrapped in a `@thunk` (potentially using a `begin`-`end` block).
+
+If there are multiple return values, their computation should almost always be wrapped in a `@thunk`.
+
+Do _not_ wrap _variables_ in a `@thunk`, wrap the _computations_ that fill those variables in `@thunk`:
+
+```julia
+# good:
+∂A = @thunk(foo(x))
+return ∂A
+
+# bad:
+∂A = foo(x)
+return @thunk(∂A)
+```
+In the bad example `foo(x)` gets computed eagerly, and all that the thunk is doing is wrapping the already calculated result in a function that returns it.
+
+### Be careful with using `adjoint` when you mean `transpose`
+
+Remember for complex numbers `a'` (i.e. `adjoint(a)`) takes the complex conjugate.
+Instead you probably want `transpose(a)`.
+
+While there are arguments that for reverse-mode taking the adjoint is correct, it is not currently the behavior of ChainRules to do so.
+Feel free to open an issue to discuss it.
+
+### Code Style
+
+Use named local functions for the `pushforward`/`pullback`:
+
+```julia
+# good:
+function frule(::typeof(foo), x)
+    Y = foo(x)
+    function foo_pushforward(_, ẋ)
+        return bar(ẋ)
+    end
+    return Y, foo_pushforward
+end
+
+# bad:
+function frule(::typeof(foo), x)
+    return foo(x), (_, ẋ) -> bar(ẋ)
+end
+```
+
+While this is more verbose, it ensures that if an error is thrown during the `pullback`/`pushforward` the [`gensym`](https://docs.julialang.org/en/v1/base/base/#Base.gensym) name of the local function will include the name you gave it.
+This makes it a lot simpler to debug from the stacktrace.
+
+### Write tests
+
+There are fairly decent tools for writing tests based on [FiniteDifferences.jl](https://github.com/JuliaDiff/FiniteDifferences.jl).
+They are in [`tests/test_utils.jl`](https://github.com/JuliaDiff/ChainRules.jl/blob/master/test/test_util.jl).
+Take a look at existing test and you should see how to do stuff.
+
+!!! warning
+    Use finite differencing to test derivatives.
+    Don't use analytical derivations for derivatives in the tests!
+    Since the rules are analytic expressions, re-writing those same expressions in the tests
+    can not be an effective way to test, and will give misleading test coverage.
+
+### CAS systems are your friends.
+
+It is very easy to check gradients or derivatives with a computer algebra system (CAS) like [WolframAlpha](https://www.wolframalpha.com/input/?i=gradient+atan2%28x%2Cy%29).
+
+------------------------------------------
+
+## FAQ
+
+### What is up with the different symbols?
+
+ - `Δx` is the input to a propagator, (i.e a _seed_ for a _pullback_; or a _perturbation_ for a _pushforward_)
+ - `∂x` is the output of a propagator
```

Comment:
If this is convention it should go above before talking about `Δself` and `∂self`

Reply:
This is the FAQ.
it is a ummary of the conventions.
They are not not defined in place